lib/js/urweb.js

```diff
@@ -27,18 +27,2882 @@ function le(x, y) { return x <= y; }
+    if (cp == 42874) return true;
+    if (cp == 42876) return true;
+    if (cp == 42879) return true;
+    if (cp == 42881) return true;
+    if (cp == 42883) return true;
+    if (cp == 42885) return true;
+    if (cp == 42887) return true;
+    if (cp == 42892) return true;
+    if (cp == 42894) return true;
+    if (cp == 42897) return true;
+    if (cp >= 42899 && cp <= 42901) return true;
+    if (cp == 42903) return true;
+    if (cp == 42905) return true;
+    if (cp == 42907) return true;
+    if (cp == 42909) return true;
+    if (cp == 42911) return true;
+    if (cp == 42913) return true;
+    if (cp == 42915) return true;
+    if (cp == 42917) return true;
+    if (cp == 42919) return true;
+    if (cp == 42921) return true;
+    if (cp == 42933) return true;
+    if (cp == 42935) return true;
+    if (cp >= 43000 && cp <= 43002) return true;
+    if (cp >= 43824 && cp <= 43866) return true;
+    if (cp >= 43868 && cp <= 43877) return true;
+    if (cp >= 43888 && cp <= 43967) return true;
+    if (cp >= 64256 && cp <= 64262) return true;
+    if (cp >= 64275 && cp <= 64279) return true;
+    if (cp >= 65345 && cp <= 65370) return true;
+    if (cp >= 66600 && cp <= 66639) return true;
+    if (cp >= 66776 && cp <= 66811) return true;
+    if (cp >= 68800 && cp <= 68850) return true;
+    if (cp >= 71872 && cp <= 71903) return true;
+    if (cp >= 119834 && cp <= 119859) return true;
+    if (cp >= 119886 && cp <= 119892) return true;
+    if (cp >= 119894 && cp <= 119911) return true;
+    if (cp >= 119938 && cp <= 119963) return true;
+    if (cp >= 119990 && cp <= 119993) return true;
+    if (cp == 119995) return true;
+    if (cp >= 119997 && cp <= 120003) return true;
+    if (cp >= 120005 && cp <= 120015) return true;
+    if (cp >= 120042 && cp <= 120067) return true;
+    if (cp >= 120094 && cp <= 120119) return true;
+    if (cp >= 120146 && cp <= 120171) return true;
+    if (cp >= 120198 && cp <= 120223) return true;
+    if (cp >= 120250 && cp <= 120275) return true;
+    if (cp >= 120302 && cp <= 120327) return true;
+    if (cp >= 120354 && cp <= 120379) return true;
+    if (cp >= 120406 && cp <= 120431) return true;
+    if (cp >= 120458 && cp <= 120485) return true;
+    if (cp >= 120514 && cp <= 120538) return true;
+    if (cp >= 120540 && cp <= 120545) return true;
+    if (cp >= 120572 && cp <= 120596) return true;
+    if (cp >= 120598 && cp <= 120603) return true;
+    if (cp >= 120630 && cp <= 120654) return true;
+    if (cp >= 120656 && cp <= 120661) return true;
+    if (cp >= 120688 && cp <= 120712) return true;
+    if (cp >= 120714 && cp <= 120719) return true;
+    if (cp >= 120746 && cp <= 120770) return true;
+    if (cp >= 120772 && cp <= 120777) return true;
+    if (cp == 120779) return true;
+    if (cp >= 125218 && cp <= 125251) return true;
+
+    return false;
+}
+
+function isUpper(c) {
+    var cp = ord(c);
+
+    if (cp >= 65 && cp <= 90) return true;
```

Comment:
The two functions isUpper and isLower are really big. What did you use for reference? Also, do you think it's possible to rewrite them so as to make them smaller? E.g. we could introduce a lookup table (code point -> some of its properties) and keep it somewhere. If that table could be simpler to modify as Unicode standard changes, it would be a win.

(If you don't feel like touching it, then I'd volunteer to make this change, if you don't mind.)

Reply:
To come up with these conditions I made an app that outputted every codepoint for which the libicu correspondent function would return true; then I made a silly c# app that "joined" sequences of chars into those `(cp >= min && cp <= max)`. I considered using a lookup but I think the file would get bigger since we dont have these "ranges" (ie, we will have `lookup[65] = true; lookup[66]=true; ... etc` I assume?).

One low hanging fruit of sorts in this approach would be to detect some cases where the sequence of codepoints obeys some logic like `(cp >= min && cp <= max && (cp % 2) == 1)` like I saw in a single instance. As of now, since these codepoints are not sequential, I'm making an individual codepoint check for each codepoint. But not sure if the size saved is worth the "hassle" (ie, in practice, I don't know how many of these cases exist).

One suggestion would be to split these functions into a separate file and if the compiler found that the function was used on the client side, only then it would be merged into the app.js. Not sure how doable or if Chlipala would agree with such an approach.

One thing I considered and ended up not doing was to do some sort of "explicit" bissect (not sure what to call it 🙃 ) on the codepoint, to minimize the number of conditions evaluated, or to add an explicit "bail out condition" at the end of a "range test". This won't, however, help with the file size. What I mean is instead of:

``` javascript
if (cp >= 30 && cp <= 70) return true;
// lots and lots of ifs, all at 1st level
if (cp >= 1234567 && cp <= 1234587 ) return true;
// more ifs
```
"bissect"
``` javascript
if (cp < 500000)
{
   if (cp >= 30 && cp <= 70) return true;
   // lots and lots of ifs
}
else 
{
   if (cp >= 1234567 && cp <= 1234587 ) return true;
   // more ifs
}
```

"bail out condition"
``` javascript
if (cp >= 30 && cp <= 70) return true;
if (cp <= 70) return false;
// lots and lots of ifs
if (cp >= 1234567 && cp <= 1234587 ) return true;
if (cp <= 1234587 ) return false;
// more ifs
```

Please feel free to send any patches. I'm very much interested in having this working on the official release 😃 . Not sure if i have to do something to enable you do patch my fork, or if you even need to patch my fork.